File: packages/shared-types/src/types/event.ts
import type { UUID, ISO8601 } from './domain';

export type EventStatus = 'published' | 'processing' | 'completed' | 'failed' | 'retrying';

export type EventPriority = 'critical' | 'high' | 'normal' | 'low';

export type EventCategory = 'memory' | 'agent' | 'auth' | 'billing' | 'system' | 'sync' | 'user' | 'integration';

export interface Event {
  id: UUID;
  type: string;
  source: string;
  category: EventCategory;
  status: EventStatus;
  priority: EventPriority;
  correlationId: UUID;
  causationId?: UUID;
  tenantId: UUID;
  userId?: UUID;
  payload: Record<string, unknown>;
  metadata: EventMetadata;
  createdAt: ISO8601;
  publishedAt?: ISO8601;
  retryCount: number;
  maxRetries: number;
}

export interface EventMetadata {
  version: number;
  schema: string;
  producer: string;
  timestamp: ISO8601;
  traceId: string;
  spanId: string;
}

export interface EventSubscription {
  id: UUID;
  eventType: string;
  handlerId: UUID;
  handlerType: 'service' | 'agent' | 'webhook' | 'function';
  config: SubscriptionConfig;
  filters?: Record<string, unknown>;
  enabled: boolean;
  createdAt: ISO8601;
}

export interface SubscriptionConfig {
  batchSize?: number;
  maxRetries: number;
  timeoutMs: number;
  deadLetter: boolean;
}

export interface DeadLetterEvent {
  id: UUID;
  originalEventId: UUID;
  error: string;
  errorCount: number;
  lastErrorAt: ISO8601;
  payload: Record<string, unknown>;
}

File: apps/api/src/api/schemas/event.py
import uuid
from datetime import datetime
from typing import Any

from pydantic import BaseModel, Field


class EventPublish(BaseModel):
    type: str = Field(..., min_length=1)
    source: str = Field(..., min_length=1)
    category: str = Field(..., min_length=1)
    payload: dict[str, Any] = Field(default_factory=dict)
    priority: str = "NORMAL"
    correlation_id: str | None = None
    workspace_id: str | None = Field(None, description="Workspace to scope event")


class EventResponse(BaseModel):
    id: uuid.UUID
    type: str
    source: str
    category: str
    status: str
    priority: str
    correlation_id: uuid.UUID
    causation_id: uuid.UUID | None = None
    payload: dict[str, Any]
    metadata: dict[str, Any] = Field(validation_alias="metadata_")
    tenant_id: uuid.UUID | None = None
    user_id: uuid.UUID | None = None
    workspace_id: uuid.UUID | None = None
    retry_count: int = 0
    max_retries: int = 3
    created_at: datetime
    published_at: datetime | None = None

    model_config = {"from_attributes": True}


class SubscriptionCreate(BaseModel):
    event_type: str = Field(..., min_length=1)
    handler_id: str = Field(..., min_length=1)
    handler_type: str = "service"
    config: dict[str, Any] = Field(default_factory=dict)
    filters: dict[str, Any] | None = None


class SubscriptionResponse(BaseModel):
    id: uuid.UUID
    event_type: str
    handler_id: uuid.UUID
    handler_type: str
    config: dict[str, Any]
    filters: dict[str, Any] | None = None
    enabled: bool = True
    created_at: datetime

    model_config = {"from_attributes": True}
